packages/x-date-pickers/src/internals/hooks/usePickerState.ts

```diff
@@ -69,95 +115,109 @@ export const usePickerState = <TInput, TDateValue>(
+      dispatch({ type: params.type, value: params.value });
 
-  // Mobile keyboard view is a special case.
-  // When it's open picker should work like closed, because we are just showing text field
-  const [isMobileKeyboardViewOpen, setMobileKeyboardViewOpen] = React.useState(false);
-
-  const acceptDate = React.useCallback(
-    (acceptedDate: TDateValue, needClosePicker: boolean) => {
-      onChange(acceptedDate);
+      if (!params.internal) {
+        onChange(params.value);
+      }
 
-      if (needClosePicker) {
+      if (params.closePicker) {
         setIsOpen(false);
-        setInitialDate(acceptedDate);
         if (onAccept) {
-          onAccept(acceptedDate);
+          onAccept(params.value);
         }
       }
     },
     [onAccept, onChange, setIsOpen],
   );
 
+  // Set the draft and committed date to equal the new prop value.
+  if (!valueManager.areValuesEqual(utils, dateState.committed, parsedDateValue)) {
+    setDate({ type: 'commit', value: parsedDateValue, internal: true });
+  }
+
+  // Mobile keyboard view is a special case.
+  // When it's open picker should work like closed, because we are just showing text field
+  const [isMobileKeyboardViewOpen, setMobileKeyboardViewOpen] = React.useState(false);
+
+  const shouldCloseOnSelect = !(disableCloseOnSelect ?? wrapperVariant === 'mobile');
+
   const wrapperProps = React.useMemo(
     () => ({
       open: isOpen,
-      onClear: () => acceptDate(valueManager.emptyValue, true),
-      onAccept: () => acceptDate(draftState.draft, true),
-      onDismiss: () => acceptDate(initialDate, true),
+      onClear: () => {
+        // Reset all date in state to the empty value and close picker
+        setDate({ value: valueManager.emptyValue, type: 'accept', closePicker: true });
+      },
+      onAccept: () => {
+        // Set all date in state to equal the current draft value and close picker
+        setDate({ value: dateState.draft, type: 'accept', closePicker: true });
+      },
+      onDismiss: () => {
+        if (shouldCloseOnSelect) {
+          // Set all dates in state to equal the last accepted date
+          // e.g. Reset the state to the last accepted value
+          setDate({ value: dateState.resetTarget, type: 'accept', closePicker: true });
+        } else {
+          // Set all dates in state to equal the last committed date
+          // e.g. Reset the state to the last committed value
+          setDate({ value: dateState.committed, type: 'accept', closePicker: true });
+        }
+      },
       onSetToday: () => {
-        const now = utils.date() as TDateValue;
-        dispatch({ type: 'update', payload: now });
-        acceptDate(now, !disableCloseOnSelect);
+        if (shouldCloseOnSelect) {
+          setDate({ value: utils.date()!, type: 'accept', closePicker: true });
+        } else {
+          setDate({ value: utils.date()!, type: 'commit' });
```

Comment:
The previous version did not took the default version of `disableCloseOnSelect` into account (see `pickerProps.onDateChange` where we do use this default version).

I don't know what should be the behavior here
Should the "Today" button behave like "Accept" (ie: close even if disableCloseOnSelect = true) or should it behave like a regular date selection (ie: only close if disableCloseOnSelect = false).

@BGehrels what do you think ?

Reply:
Phu, depends on the primary use case (we are not using it):

- If it's to select a date, i would close/switch to the TimePicker when it is clicked (What is your preferred Delivery date? Today, 5:30! - 3 clicks)
- If it's to quickly get back to the current calendar page i would leave it open (Preselection is 1987, today is 2022, i don't want to click "next year" 35 times, get me there fast, please).

Do you have user research/tracking insights which of those two is more relevant? My gut feeling would say the first one, so immediately closing the picker would make most sense - especially, since you can always reopen it in case you wanted the later case.

Reply:
> Do you have user research/tracking insights which of those two is more relevant?

Not yet, I think I will try to propose a behavior a clear as possible in most scenarios and to list clearly those where there is a debate (the one you pointed out about the rest when dismissing a multi-views picker for instance).